Add -c flag to count the track number in a 3digit wide playlist order

thescooby opened the pull request:
Wanted to have the following features:
-   The files shall be sorted into folders like \<artist\> - \<album\>
-   The files shall have a file name prefix (e.g. 012_*) that counts up per playlist track. 
    This way the files are stored in the same order on disk as they are in the playlist.

Bleuzen commented on spotrec.py:
```diff
@@ -409,8 +419,20 @@ class FFmpeg:
         self.pulse_input = _pa_recording_sink_name + ".monitor"
 
+        if _use_pl_track_counter:
+            # expect incoming file name format like '05 - artist - song' 
```
This may lead to errors since users can change the filename format with the --filename-pattern option

Bleuzen commented on spotrec.py:
```diff
@@ -409,8 +419,20 @@ class FFmpeg:
+
         self.pulse_input = _pa_recording_sink_name + ".monitor"
 
+        if _use_pl_track_counter:
+            # expect incoming file name format like '05 - artist - song' 
+            # test if fourth char is a '-'
+            if filename[3] == '-':
+                filename = filename[5:]
```
This may lead to errors since users can change the filename format with the --filename-pattern option

thescooby replied:
Yes, you are right. Will change it.

Bleuzen commented:
Hi, thanks for your PR!
At the first look, these came to my:

1.)
The playlist track counter is dependent on the default filename pattern (locks users out from using the --filename-pattern option).

2.)
The sort_in_folders option may be unnecessary. Here is an alternative, cleaner idea that came to my mind: implement it into the existing --filename-pattern option. So using it could look like this:

`./spotrec.py [...] --filename-pattern "{artist} - {album}/{trackNumber} - {title}"`

Bleuzen commented:
Forgot this last night, but now that I think of it:
The playlist-counter should also be implemented in the existing --filename-pattern option, allowing for more flexible filename/-paths. This way, no new command line option is needed, keeping it simpler and not forcing any format (where the playlist-count is) in the filename.

Bleuzen commented on spotrec.py:
```diff
@@ -49,6 +51,7 @@
 _shell_executable = "/bin/bash"  # Default: "/bin/sh"
 _shell_encoding = "utf-8"
 _ffmpeg_executable = "ffmpeg"  # Example: "/usr/bin/ffmpeg"
+_pl_track_counter = 1
```
This variable is placed under the "# Hard-coded settings" section. This is meant for constants. It should be moved down to "# Variables that change during runtime"

thescooby commented:
My changes:
- removed '-a' flag, did not fully understand the filename-pattern feature completely at first, but feature had issues with folders specified
- fixed this folder handling

Works now with e.g.:
spotrec.py [..] -p "{artist} - {album}/{trackNumber} - {title}"
and even with
spotrec.py [..] -p "{artist} - {album}/2020/{trackNumber} - {artist} - {title}"

Cool. ;-)
Thanks for mentioning this, it is a much cleaner design now.

The playlist numbering suffered from having too many on_playing_uri_changed() events. Strange. Therefore I put the playlist counter increasing into stop_blocking(). But even here I get sometimes very small files (10-30kB) and then spotrec starts again by a file that sometimes is already completely downloaded. Then the playlist counter jumps of course. I couldn't find the bug here this time, e.g.
001 - Dermot Kennedy - Outnumbered.flac (23 MB)
002 - Nico Santos - Play With Fire.flac (23,1 MB)
003 - Imagine Dragons - Bad Liar.flac (26 kB)
004 - Nico Santos - Play With Fire.flac (23,1 MB)
005 - Imagine Dragons - Bad Liar.flac ()

Bleuzen commented:
Somehow GitHub doesn't like the force pushes. Looking at the summary:
https://github.com/Bleuzen/SpotRec/pull/13/files
for example the pl_track_counter variable is still under the hard coded ones, even while you moved it.
Don't know yet if more is broken.

Bleuzen commented:
I like the new implementation of the playlist track counter and the fact that it just replaces the trackNumber, which is otherwise the album trackNumber.
Description of the command line option still need to be changed to reflect that.

thescooby commented:
https://github.com/Bleuzen/SpotRec/pull/13#issuecomment-780550277: I can not see pl_track_counter still under "hard coded", just disable the folding and then it is under the runtime variables. Regarding force-push: I looked again through the file, looks good so far. But maybe the "github folding" let the file look strange.

Bleuzen commented on spotrec.py:
```diff
@@ -39,6 +39,7 @@
 _filename_pattern = "{trackNumber} - {artist} - {title}"
 _tmp_file = True
 _underscored_filenames = False
+_use_intern_track_counter = False
```
minor one: meaning of "intern": "a student or trainee who works, sometimes without pay, at a trade or occupation in order to gain work experience."

guess "internal" is the correct term :D

thescooby replied:
Oh yes, thanks. I am german and will never learn it... :-D

thescooby replied:
Done with latest commit.

Bleuzen replied:
Heißt bei uns btw. sowas wie Praktikant, was hier keinen Sinn ergibt. An sich hab ich nichts gegen Abkürzungen, aber sie müssen eindeutig sein :D

Bleuzen commented on spotrec.py:
```diff
@@ -184,7 +189,7 @@ class Spotify:
 
     def __init__(self):
-        dbus.mainloop.glib.DBusGMainLoop(set_as_default=True)
+        self.glibloop = dbus.mainloop.glib.DBusGMainLoop(set_as_default=True)
```
Is there a reason the glibloop is stored in a variable now or can this be removed?

thescooby replied:
Yes, later on you reference it in quit_glib_loop(). My linter said this variable is not available.

Bleuzen replied:
The variable self.glibloop is set here:
https://github.com/Bleuzen/SpotRec/blob/a9706601b96f6871576b872c4f21974b67cbd487/spotrec.py#L216

In quit_glib_loop(), this is called:
`self.glibloop.quit()`

This works because the GLib.MainLoop (set in DBusListenerThread) has a quit function.

The DBusGMainLoop does not even have a quit() function, so this wont work.

Setting self.glibloop here sets it to the wrong value.
It still works, because it is overwritten shortly after, but it is still wrong.

An alternative to make the linter happy should be found.

thescooby replied:
Ah, sorry. Will fix.

thescooby replied:
Updated.

Bleuzen commented on spotrec.py:
```diff
@@ -250,10 +255,10 @@ def get_track(self, metadata):
         ret = str(filename_pattern.format(
             artist=self.metadata_artist,
-            album=self.metadata_artist,
+            album=self.metadata_album,
```
nice catch, thanks

Bleuzen commented:
Thank you, looks good now!
Think this can be merged now. I will do some testing and if everything goes well this will land in a new release soon.

Bleuzen commented:
Found one problem left: the internal_track_counter increment happens too late. So for example the first two songs recorded are prefixed with "001".

This is because stopping the recording does not happen immediately, it happens in this order:
- spotrec updates metadata on start, putting in 001 as trackNumber
- The first tracks plays and finishes
- spotrec detects a song change-
- spotrec updates the metadata
- the old recording is stopped (and this is the place where you currently increment the counter)

Incrementing the counter needs to happen before or while the metadata gets update. Currently it happens after, which does not update the counter in time.

thescooby commented:
> Found one problem left: the internal_track_counter increment happens too late. So for example the first two songs recorded are prefixed with "001".
> 
> This is because stopping the recording does not happen immediately, it happens in this order:
> 
>     * spotrec updates metadata on start, putting in 001 as trackNumber
> 
>     * The first track plays
> 
>     * spotrec uses current count (001) for the recording
> 
>     * The first track finishes
> 
>     * spotrec detects a song change
> 
>     * spotrec updates the metadata
> 
>     * spotrec starts the new recording (of the second song) (track counter is still 001 here)
> 
>     * the old recording (of the first song) is stopped (and this is the place where you currently increment the counter)
> 
> 
> For now I put the increment code here:
> [1224986](https://github.com/Bleuzen/SpotRec/commit/12249865c545999dde3e9941ca9d497de517eae2)
> 
> Which is not optimal as it counts every track change (also the ones not recorded), but at least it reliably increments at time and puts the output files in order.

thescooby commented:
Yes, because of getting triggered many times I put it in stop_blocking(). But you are right at least it counts up, I got really big jumps in the counter like e.g. 001 and the next was 016 etc.

Bleuzen commented:
@thescooby Your changes are now released! :)
https://github.com/Bleuzen/SpotRec/releases/tag/v0.14.0

FYI I made some more changes, for your feature: improved stability of the track counter (https://github.com/Bleuzen/SpotRec/commit/bdd1c2a4cd8ea5452c651359fe7aaaaca14463fb) and added example in the command line option help text
